examples/algorithm_progression.py

```diff
@@ -0,0 +1,177 @@
+
+    source = large_image.open(input_filename)
+    param_space_combos = itertools.product(*input_params.values())
+    combos_uniques = itertools.product(
+        *[p for p in input_params.values() if len(p) > 1]
+    )
+
+    print(f'Beginning {len(param_space_combos)} runs on {MAX_THREADS} threads...')
+    with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
+        for result in executor.map(
+            lambda combo, unique: save_result(
+                apply_algorithm(algorithm, source, combo),
+                *unique,
+                output_dir,
+                algorithm_name,
+            ),
+            param_space_combos,
+            enumerate(combos_uniques),
+        ):
+            yaml_dict['sources'].append(result)
+
+    with open(Path(output_dir, 'results.yml'), 'w') as f:
+        yaml.dump(
+            yaml_dict,
+            f,
+            default_flow_style=False,
+            sort_keys=False,
+        )
+
+
+def get_input_params_from_user(algorithm, sig):
+    print(f'\nProvide parameter space for algorithm {algorithm.__name__}.')
+    print(
+        'Hit enter to accept default for every run, '
+        'provide a single value to replace the default for every run.'
+    )
+    print(
+        'Provide two values separated by a comma '
+        'to specify a range to cover across all runs.'
+    )
+    input_params = {}
+    for param in sig.parameters.values():
+        if param.name != 'data':
+            user_input = input(f'{param.name} ({param.default}): ')
+            if len(user_input.split(',')) == 2:
+                user_range = [float(i.strip()) for i in user_input.split(',')]
+                num_values = int(
+                    input(
+                        f'Enter a number of values to generate in the range {user_range}:'
+                    )
+                )
+                input_params[param.name] = np.linspace(
+                    user_range[0], user_range[1], num=num_values
+                )
+            elif len(user_input):
+                input_params[param.name] = [user_input]
+            else:
+                input_params[param.name] = [param.default]
+    return input_params
+
+
+if __name__ == '__main__':
+    args = argparser.parse_args()
+
+    algorithm_code = args.algorithm_code
+    input_filename = args.input_filename
+    output_dir = args.output_dir
+
+    if not Path(input_filename).exists():
+        raise ValueError(f'Cannot locate file {input_filename}.')
+
+    algorithm = algorithms.ALGORITHM_CODES[algorithm_code]
+    sig = inspect.signature(algorithm)
+
+    if not output_dir:
+        output_dir = f'{algorithm.__name__}_output'
+        i = 0
+        while Path(output_dir).exists():
+            i += 1
+            output_dir = f'{algorithm.__name__}_output_{i}'
+        os.mkdir(output_dir)
+    else:
+        if not Path(output_dir).exists():
+            os.mkdir(output_dir)
+
+    user_provided = input(
```

Comment:
Having user input is fine as a feature, but tedious for reproducible testing.  I'd rather be able to say `--param=hue_value,0,1,250,open --param=hue_width=0.05,0.25,9`.  There is a lot we could include: open-ended/close-ended interval, number of steps versus step size, axis name (and we now can support arbitrary axes), etc., but I think it would suffice that if any algorithm parameters were added on the command line we'd skip asking the user.